### PR TITLE
Remove VP constants on deactivation

### LINKED ISSUE #1021: VP constants are present in wp-config.php after plugin deactivation
Steps to reproduce: 
- Install WordPress
- in `wp-config.php` replace line`/* That's all, stop editing! Happy blogging. */` with custom constant e.g. `define('my_constant', true);`
- activate and deactivate VersionPress

Lines below remains in file.
```

/** Absolute path to the WordPress directory. */
define('VP_ENVIRONMENT', 'default');

```

Related issue #1020

### PR DESCRIPTION
**Pull request by octopuss**
Resolves #1021 .

VersionPress constants are removed on uninstall. Also performed refactor of public constants (`VP_` prefixed)

Reviewers:

- [x] @JanVoracek 

Thank you

### CODE REVIEW

File: plugins/versionpress/src/Initialization/WpConfigSplitter.php [hunk 1]
```diff
@@ -35,9 +35,10 @@ public static function split($wpConfigPath, $commonConfigName = self::COMMON_CON
         $commonConfigLines = is_file($commonConfigPath)
             ? file($commonConfigPath)
             : file(__DIR__ . '/wp-config.common.tpl.php');
-
+        $constantsClass = new ReflectionClass(PublicConstants::class);
+        $vpPublicConstants = array_values($constantsClass->getConstants());
         // https://regex101.com/r/zD3mJ4/2
-        $constantsForRegex = join('|', self::$constantsForExtraction);
+        $constantsForRegex = join('|', array_merge(self::$wpConstantsForExtraction, $vpPublicConstants));
```
Comment (JanVoracek) [inline plugins/versionpress/src/Initialization/WpConfigSplitter.php]: We definitely don't want to move all public constants to the common config. For example `VP_ENVIRONMENT` or `VP_GIT_BINARY` are environment-specific. Thus, they need to stay in the `wp-config.php`.

**Comment by JanVoracek** [comment]
I'm still not sure about using `constant` function to get the value of the constant. I need some time to think about it.

**Comment by octopuss** [comment]
Ok so i will revert it and we will use just list of VP constants that
should be removed. I think that it will be cleaner, more straight forward
solution without big drawback.

On Wed, May 11, 2016 at 4:18 PM, Jan Voráček <notifications@github.com>
wrote:

> I'm still not sure about using constant function to get the value of the
> constant. I need some time to think about it.
>
> —
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub
> <https://github.com/versionpress/versionpress/pull/1042#issuecomment-218473333>
>



-- 
S pozdravom a pekný deň praje,
Ing. Ivan Doležal
http://www.octopuss.sk

File: plugins/versionpress/src/Utils/WpConfigEditor.php [hunk 2]
```diff
@@ -38,6 +45,20 @@ public function updateConfigConstant($constantName, $value, $usePlainValue = fal
...
 
     /**
+     * Removes VersionPress public constants from config files.
+     * @param array $configFiles List of config file paths from which constants should be removed.
+     */
+    public static function removeVersionPressConstants($configFiles)
+    {
+        foreach ($configFiles as $configFile) {
+            $constantsForRegex = join('|', self::$vpPublicConstants);
+            $defineRegexPattern = "/(define\\s*\\(\\s*['\"]($constantsForRegex)['\"]\\s*,.*\\)\\s*;)/m";
```
Comment (JanVoracek) [inline plugins/versionpress/src/Utils/WpConfigEditor.php]: Can you please add a link to regex101?

File: plugins/versionpress/setup-hooks.php [hunk 3]
```diff
@@ -659,6 +659,11 @@ function vp_admin_post_confirm_deactivation()
         FileSystem::remove(VERSIONPRESS_ACTIVATION_FILE);
     }
 
+    $wpConfigPath = \VersionPress\Utils\WordPressMissingFunctions::getWpConfigPath();
```
Comment (JanVoracek) [inline plugins/versionpress/setup-hooks.php]: This should be in the uninstallation script. We don't want to remove the constants when a user only deactivates VersionPress and activates it again (for example during an update).

**Comment by JanVoracek** [comment]
It was merged into wrong branch (`master` instead of `3.x`). I re-merged it in 17bef5bd1f0cefe41191e476906d85dad61b064e.

**Pull request merged**